A crash occurred if while load saved survey press back button

FATAL EXCEPTION: AsyncTask #2
                                                                                Process: org.eyeseetea.surveillance_mm_bb, PID: 4451
                                                                                java.lang.RuntimeException: An error occurred while executing doInBackground()
                                                                                    at android.os.AsyncTask$3.done(AsyncTask.java:309)
                                                                                    at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
                                                                                    at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
                                                                                    at java.util.concurrent.FutureTask.run(FutureTask.java:242)
                                                                                    at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
                                                                                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
                                                                                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
                                                                                    at java.lang.Thread.run(Thread.java:818)
                                                                                 Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.Context android.app.Activity.getApplicationContext()' on a null object reference
                                                                                    at org.eyeseetea.malariacare.fragments.SurveyFragment.prepareTab(SurveyFragment.java:206)
                                                                                    at org.eyeseetea.malariacare.fragments.SurveyFragment.access$500(SurveyFragment.java:63)
                                                                                    at org.eyeseetea.malariacare.fragments.SurveyFragment$AsyncReloadAdaptersAndChangeTab.doInBackground(SurveyFragment.java:305)
                                                                                    at org.eyeseetea.malariacare.fragments.SurveyFragment$AsyncReloadAdaptersAndChangeTab.doInBackground(SurveyFragment.java:280)
                                                                                    at android.os.AsyncTask$2.call(AsyncTask.java:295)
                                                                                    at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                                    at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234) 
                                                                                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
                                                                                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
                                                                                    at java.lang.Thread.run(Thread.java:818)

--- a/app/src/main/java/org/eyeseetea/malariacare/DashboardActivity.java
+++ b/app/src/main/java/org/eyeseetea/malariacare/DashboardActivity.java
@@ -454,6 +454,11 @@ public class DashboardActivity extends BaseActivity {
      * It is called when the user press back in a surveyFragment
      */
     private void onSurveyBackPressed() {
+        //// FIXME: 09/03/2017 Refactor. If the survey is loading the survey should be closed at the end
+        if(Session.isIsLoadingSurvey()){
+            Session.setShouldPressBackOnLoadSurvey(true);
+            return;
+        }
         Log.d(TAG, "onBackPressed");
         Survey survey = Session.getMalariaSurvey();
         if (!survey.isSent()) {
@@ -490,26 +495,31 @@ public class DashboardActivity extends BaseActivity {
      * After that, loads the Assess fragment(DashboardUnSentFragment) in the Assess tab.
      */
     public void closeSurveyFragment() {
-        boolean isSent = false;
-        isReadOnly = false;
-        isLoadingReview = false;
-        android.support.v7.app.ActionBar actionBar = this.getSupportActionBar();
-        LayoutUtils.setDashboardActionBar(actionBar);
-        tabHost.getTabWidget().setVisibility(View.VISIBLE);
-        ScoreRegister.clear();
-        if (Session.getMalariaSurvey() != null) {
-            isSent = Session.getMalariaSurvey().isSent();
-        }
-        if (isBackPressed) {
-            beforeExit();
-        }
-        surveyFragment.unregisterFragmentReceiver();
-        if (isSent) {
-            tabHost.setCurrentTabByTag(getResources().getString(R.string.tab_tag_improve));
-            initAssess();
-        } else {
-            initAssess();
-            unsentFragment.reloadData();
+        //FIXME: 09/03/2017  Refactor: This is used to prevent multiple open and close surveys crash
+        //The survey only can be closed when is load
+        if(!Session.isIsLoadingSurvey()) {
+            Session.setShouldPressBackOnLoadSurvey(false);
+            boolean isSent = false;
+            isReadOnly = false;
+            isLoadingReview = false;
+            android.support.v7.app.ActionBar actionBar = this.getSupportActionBar();
+            LayoutUtils.setDashboardActionBar(actionBar);
+            tabHost.getTabWidget().setVisibility(View.VISIBLE);
+            ScoreRegister.clear();
+            if (Session.getMalariaSurvey() != null) {
+                isSent = Session.getMalariaSurvey().isSent();
+            }
+            if (isBackPressed) {
+                beforeExit();
+            }
+            surveyFragment.unregisterFragmentReceiver();
+            if (isSent) {
+                tabHost.setCurrentTabByTag(getResources().getString(R.string.tab_tag_improve));
+                initAssess();
+            } else {
+                initAssess();
+                unsentFragment.reloadData();
+            }
         }
     }
 

--- a/app/src/main/java/org/eyeseetea/malariacare/data/database/utils/Session.java
+++ b/app/src/main/java/org/eyeseetea/malariacare/data/database/utils/Session.java
@@ -73,6 +73,13 @@ public class Session {
      */
     private static PhoneMetaData phoneMetaData;
     /**
+     * The exitOfSurvey
+     */
+    //FIXME: 09/03/2017    This variable should disappear instance of a refactor in loading of a new survey
+    private static boolean isLoadingSurvey;
+    //FIXME: 09/03/2017    This variable should disappear instance of a refactor in loading of a new survey
+    private static boolean shouldPressBackOnLoadSurvey;
+    /**
      * The maximum total of questions in programm
      */
     private static int maxTotalQuestions;
@@ -81,6 +88,22 @@ public class Session {
      * Map that holds non serializable results from services
      */
     private static Map<String, Object> serviceValues = new HashMap<>();
+
+    public static boolean shouldPressBackOnLoadSurvey() {
+        return shouldPressBackOnLoadSurvey;
+    }
+
+    public static void setShouldPressBackOnLoadSurvey(boolean shouldPressBackOnLoadSurvey) {
+        Session.shouldPressBackOnLoadSurvey = shouldPressBackOnLoadSurvey;
+    }
+
+    public static boolean isIsLoadingSurvey() {
+        return isLoadingSurvey;
+    }
+
+    public static void setIsLoadingSurvey(boolean value) {
+        isLoadingSurvey = value;
+    }
 
     public static Survey getMalariaSurvey() {
         return malariaSurvey;

--- a/app/src/main/java/org/eyeseetea/malariacare/fragments/SurveyFragment.java
+++ b/app/src/main/java/org/eyeseetea/malariacare/fragments/SurveyFragment.java
@@ -203,9 +203,6 @@ public class SurveyFragment extends Fragment implements IDashboardFragment {
      * Prepares the selected tab to be shown
      */
     private View prepareTab(Tab selectedTab) {
-        if(getActivity()==null || getActivity().getApplicationContext()==null) {
-            return null;
-        }
         LayoutInflater inflater = LayoutInflater.from(getActivity().getApplicationContext());
 
         if (selectedTab.isCompositeScore()) {
@@ -313,8 +310,7 @@ public class SurveyFragment extends Fragment implements IDashboardFragment {
         @Override
         protected void onPostExecute(View viewContent) {
             super.onPostExecute(viewContent);
-            if(viewContent==null)
-                return;
+
             Tab tab = tabs.get(0);
 
             content.removeAllViews();
@@ -340,6 +336,8 @@ public class SurveyFragment extends Fragment implements IDashboardFragment {
         @Override
         public void onReceive(Context context, Intent intent) {
             Log.d(TAG, "onReceive");
+            //FIXME: 09/03/2017  Refactor: This is used to prevent multiple open and close surveys crash
+            Session.setIsLoadingSurvey(true);
             List<CompositeScore> compositeScores;
             List<Tab> tabs;
             Session.valuesLock.readLock().lock();

--- a/app/src/main/java/org/eyeseetea/malariacare/fragments/SurveyFragment.java
+++ b/app/src/main/java/org/eyeseetea/malariacare/fragments/SurveyFragment.java
@@ -203,6 +203,9 @@ public class SurveyFragment extends Fragment implements IDashboardFragment {
      * Prepares the selected tab to be shown
      */
     private View prepareTab(Tab selectedTab) {
+        if(getActivity()==null || getActivity().getApplicationContext()==null) {
+            return null;
+        }
         LayoutInflater inflater = LayoutInflater.from(getActivity().getApplicationContext());
 
         if (selectedTab.isCompositeScore()) {
@@ -310,7 +313,8 @@ public class SurveyFragment extends Fragment implements IDashboardFragment {
         @Override
         protected void onPostExecute(View viewContent) {
             super.onPostExecute(viewContent);
-
+            if(viewContent==null)
+                return;
             Tab tab = tabs.get(0);
 
             content.removeAllViews();

--- a/app/src/main/java/org/eyeseetea/malariacare/layout/adapters/survey/DynamicTabAdapter.java
+++ b/app/src/main/java/org/eyeseetea/malariacare/layout/adapters/survey/DynamicTabAdapter.java
@@ -504,6 +504,16 @@ public class DynamicTabAdapter extends BaseAdapter implements ITabAdapter {
 
         rowView.requestLayout();
         reloadingQuestionFromInvalidOption = false;
+
+        //FIXME: 09/03/2017  Refactor: This is used to prevent multiple open and close surveys crash
+        Session.setIsLoadingSurvey(false);
+        if(Session.shouldPressBackOnLoadSurvey()) {
+            DashboardActivity.dashboardActivity.runOnUiThread(new Runnable() {
+                public void run() {
+                    DashboardActivity.dashboardActivity.onBackPressed();
+                }
+            });
+        }
         return rowView;
     }
 

